### PR TITLE
Replace snapshot id with url parameters

### PR DESCRIPTION
**Pull request by bocasti**
Once again, it looks (and it got) bigger that what I had in mind when starting to move things! sorry for that :(

At the end, here are the changes that this PR would introduce:

1) A snapshot would have a #snapshotId and an #exceptionId
The #exceptionId identify a specific exception reported by SnapDump (what was previously the #id)
The #snapshotId identify a specific snapshot being registered for a given exception. Several #snapshotId can be associated to the same #exceptionId

The intention is to be able to collect several snapshots for a similar exception occuring at different point in time.
With a limit  on server side on the number of snapshot that can be registered for a same #exceptionId.
This could give information about the frequency of a given exception.
Maybe later some specific context information could be reported within each snapshot.

2) I kept the concept of #snapshotId and snapshot calls with a single path parameter.
I think it would still be handy to perform some server queries with just an id.
And anyway the server perform a lot of actions in a "flat way": 
it iterates on a memory list of snapshots, that gets reset on each addition / removal. 
I understand that all snapshots should always be "manageable" in memory, as snapdump is not intended to archive snapshots. 

3) The handler is no longer initialized with project and version objects, but just project name and version strings.
The snapshots generated by handler also point to a simple project name and version string (project and version objects are left nil).
This goes in the direction of having a handler as simple as possible, not depending from the snapdump model

4) The snapshot call would have some optional query parameters: #projectName #versionName #exceptionId
They are currently used by the #put and #head actions
Doing that way makes it easier to build dynamic url (on handler side) with possible empty values for projectName and versionName.
It also feels more flexible than requiring path parameters for any http method. 

The server decides how he wants to manage empty project and version values. He is the only one that has to deal with default project and version names.
(Currently this change has limited interest regarding version and project parameters, as these ones could already be read from meta information). 

5) The file structure on server side has changed from flat to a project dir => version dir => snapshot file structure.
Right now there is no actual benefit with this change, as most of the actions are taking advantages of the snapshot memory list.
But it feels natural without adding much complexity.
 

All the tests are green and the examples work fine.

Here are a list of plans / "useful" idea that I had in mind while working on that.
Feel free to tell me if it goes totally out of scope :)

- update the UI to take advantage of the #exceptionId concept.
I guess I would start with an extra column to group snapshots from a same exception id.
Also having the ability to remove all snapshot from a same exceptionId.

- attach a "textual" stack trace to the snapshot generated by the handler.
It would be nice to still be able to take advantage of snapshots without having the exact same image context that the reported exception.
The UI could offer to display / download stack traces for a given snapshot.

- I am not sure how to do that without making snapdump too specific, but it would nice if we could also plug and attach some "log extracts" to the snap dump generated by the handler.
In our case, we could for example have access to last 30 lines of the application log before the exception occured.
From times to times, such logs context could help understand.
And managing docker container logs is a pain :)

- A minimal seaside interface for the server image ? 
This would give a quick overview of to the current reported snapshot without having to load and open snapdump in our local images.
And such information like stack-traces or logs could already be accessible from this interface ?

### CODE REVIEW

**Comment by bocasti** [comment]
Some discussed points that should be worked on:

- max number of uploadable snapshots for a given exception should be a configuration parameter
- project and version should be required by the server calls. Using a default project and version could end up by mixing snapshots from different projects
- go further with /version/snapshot directory hierarchy. take advantage of this new structure in order to remove the memory caching and optimize the PUT requests
-  house keeping:  On remove operation the directory needs to be checked afterwards and deleted if empty
- change the snapshot format separator: The problem with a single 255 byte is that I'm not sure it can appear as second or third byte. 
ideas: TLV kind of thing, use LENGTHLENGTH|LENGTH|JSON|FUEL

**Comment by bocasti** [comment]
Here are the adjustments provided by the last commit

- Server file structure is now: Project->Version->Exception->Snapshot.snap
- Server no longer cache any snapshot in memory. It always re-build snapshots from file structure.
- 'serverName' 'versionName' and 'exceptionId' are mandatory parameters for snapshots PUT and HEAD calls
- 'serverName' 'versionName' and 'exceptionId' are optional parameters for snapshots GET and DELETE calls. Because it is handy 
 to keep being able to provide only the id, and it smooth out migration to new file structure
- I made sure that when a snapshot is removed, empty parent directories are also removed
- All tests are green. I adjusted and added some
- The current client UI seems to works correctly. I started changing it in another branch

File: source/SnapDump-Client/SDHTTPStore.class.st [hunk 1]
```diff
@@ -98,12 +112,12 @@ SDHTTPStore >> snapshotsWithVersion: aVersion [
 SDHTTPStore >> versionsOfProject: aProject [ 
 	| response |
 	response := self client
-		url: uri / #projects / aProject name / #versions;
+		url: uri / #projects / aProject name urlEncoded / #versions;
```
Comment (noha) [inline source/SnapDump-Client/SDHTTPStore.class.st:115]: Why is there a urlForSnapshot: but no #urlForProject: , #urlForVersion: they can use each other and so there is only one place for the urlEncoded

File: source/SnapDump-Server/SDSnapshotCall.class.st [hunk 2]
```diff
@@ -75,5 +130,25 @@ SDSnapshotCall >> put [
...
+	^ self store
+		snapshotWithId: snapshotId
+		project: projectName
+		version: versionName ]
+		on: NotFound
+		do: [ :err | ZnRespond signalWith: (ZnResponse notFound: request url) ]
+]
+
+{ #category : #public }
+SDSnapshotCall >> validateRequiredParameters: aCollectionOfParameterNames [
```
Comment (noha) [inline source/SnapDump-Server/SDSnapshotCall.class.st:147]: Why is this here? Does the openApi one not work?

Comment (bocasti) [reply source/SnapDump-Server/SDSnapshotCall.class.st:147]: I wanted some parameters to be required only for particular call method (#post #head)
I think openApi only allows to make parameters required for all existing call methods, and not for a specific call method, right ?
Anyway I think this will be removed while moving these parameters as path parameters

File: source/SnapDump-Client/SDHTTPStore.class.st (hunk 1 above)
Comment (bocasti) [reply source/SnapDump-Client/SDHTTPStore.class.st:115]: Right, I guess I have missed that !
